Implement updated alerting & feature activations

zathras-crypto opened the pull request:
This PR serves to add support for multiple alerts with a new structure and integrated feature activations.

Work from #169 and #175 continues here.  WIP, not ready for merge.

dexX7 commented on src/omnicore/notifications.cpp:
```diff
@@ -19,8 +20,50 @@
+            PrintToLog("Removing deleted alert (from:%s type:%d expiry:%d message:%s)\n", alert.alert_sender,
+                alert.alert_type, alert.alert_expiry, alert.alert_message);
+            it = currentOmniAlerts.erase(it);
+            uiInterface.OmniStateChanged();
```
Hm, why?

zathras-crypto replied:
See my comments here https://github.com/OmniLayer/omnicore/pull/169#discussion_r37033702 mate

dexX7 commented on src/omnicore/rules.cpp:
```diff
@@ -301,6 +308,21 @@ bool IsAllowedOutputType(int whichType, int nBlock)
 }
 
 /**
+ * Adds a pending activation to the PendingActivations vector.
```
It's currently unused, but I'm wondering, what you have in mind? Add support to lookup pending activations at runtime?

zathras-crypto replied:
Originally I was going to have a function to check pending alerts, see if one went active in the block we just processed, and perform any necessary actions.

Then while working it I found the only necessary action was the "X has gone live" notification, which is easily handled via the expiry of the "X is about to go live" notification without using the pending stuff.

So long story short, not sure if I'll use it.  Pending activations at runtime can be seen via the ```omni_getinfo``` RPC call as there will be a alert for it.

dexX7 commented on src/omnicore/rules.cpp:
```diff
@@ -317,41 +339,45 @@ bool ActivateFeature(uint16_t featureId, int activationBlock, int transactionBlo
+    PrintToLog("Feature activation of ID %d processed. %s will be enabled at block %d.\n", featureId, featureName, activationBlock);
+    AddPendingActivation(featureId, activationBlock, minClientVersion, featureName);
+    AddAlert("omnicore", ALERT_BLOCK_EXPIRY, activationBlock, strprintf("Feature %d ('%s') will go live at block %d", featureId, featureName, activationBlock));
+    if (!supported) {
```
Whenever I encounter an unsupported activation, the client freezes. To be more specific:
```
# fresh start, regtest mode with -omnialertallowsender=any
setgenerate true 101
# send some coins to a new address
omni_sendactivation mr49RtQEnNz1eukzJxxEv8hGRBu5cAjwp3 9 110 0
setgenerate true 1
# freeze ...
```
Or:
```
# fresh start, regtest mode with -omnialertallowsender=any
setgenerate true 101
# send some coins to a new address
omni_sendactivation mr49RtQEnNz1eukzJxxEv8hGRBu5cAjwp3 1 110 999999999
setgenerate true 1
# freeze ...
```
Valid activations don't cause a freeze, and go through until the end (-> actual activation -> no longer a notice after 1024 in the UI).

Restarting or reparsing doesn't help, and once the unsupported activation is processed, it ends, both via `omnicored` and `omnicore-qt`.

Log:
```
2015-08-17 00:57:17 Class C transaction detected: e7b58c7434a2a6f4ccb6a917bc28a0660e0522a246346e1964d18619ffb1bf8a parsed to fffffffe00060000006e3b9ac9ff at vout 0
2015-08-17 00:57:17 single_pkt: fffffffe00060000006e3b9ac9ff
2015-08-17 00:57:17 	------------------------------
2015-08-17 00:57:17 	         version: 65535, class C
2015-08-17 00:57:17 	            type: 65534 (Feature Activation)
2015-08-17 00:57:17 	      feature id: 6
2015-08-17 00:57:17 	activation block: 110
2015-08-17 00:57:17 	 minimum version: 999999999
2015-08-17 00:57:17 	          sender: mr5Q4dPahLvfdmyouemQHFgQeRcMP8GXx3
2015-08-17 00:57:17 	      authorized: true
2015-08-17 00:57:17 Feature activation requested (ID 6 to go active as of block: 110)
2015-08-17 00:57:17 Feature activation of ID 6 processed. Unknown feature will be enabled at block 110.
2015-08-17 00:57:17 New alert added: omnicore, 1, 110, Feature 6 ('Unknown feature') will go live at block 110
2015-08-17 00:57:17 WARNING!!! AS OF BLOCK 110 THIS CLIENT WILL BE OUT OF CONSENSUS AND WILL AUTOMATICALLY SHUTDOWN.
2015-08-17 00:57:17 New alert added: omnicore, 0, 110, Your client must be updated and will shutdown at block 110 (unsupported feature 6 ('Unknown feature') activated)

2015-08-17 00:57:17 recordTX(e7b58c7434a2a6f4ccb6a917bc28a0660e0522a246346e1964d18619ffb1bf8a, valid=YES, block= 103, type= 65534, value= 0)
2015-08-17 00:57:17 recordTX(): OK, line 3389, file: omnicore/omnicore.cpp
2015-08-17 00:57:17 devmsc=4181831959914, exodus_prev=4181831482244, exodus_delta=477670
2015-08-17 00:57:17 update_tally_map(mpexoDuSkGGqvqrkrjiFng38QPkJQVFyqv, 1=0x1, +477670, ttype=0): before=4181831482244, after=4181831959914
2015-08-17 00:57:17 devmsc for block 103: 477670, Exodus balance: 41818.31959914
```

Edit: updated the log with `-omnidebug=all`. It looks like the freeze happens, when the *new* block is processed.

dexX7 replied:
With some more debug logging directly in `CheckExpiredAlerts`, it looks like there is an endless loop:

```
2015-08-17 01:04:26 CheckExpiredAlerts: alert type: 0, alert expiry: 110, alert message: Your client must be updated and will shutdown at block 110 (unsupported feature 6 ('Unknown feature') activated)
2015-08-17 01:04:26 CheckExpiredAlerts: ALERT_FEATURE_UNSUPPORTED
2015-08-17 01:04:26 CheckExpiredAlerts: alert type: 0, alert expiry: 110, alert message: Your client must be updated and will shutdown at block 110 (unsupported feature 6 ('Unknown feature') activated)
2015-08-17 01:04:26 CheckExpiredAlerts: ALERT_FEATURE_UNSUPPORTED
2015-08-17 01:04:26 CheckExpiredAlerts: alert type: 0, alert expiry: 110, alert message: Your client must be updated and will shutdown at block 110 (unsupported feature 6 ('Unknown feature') activated)
...
```

zathras-crypto replied:
Thanks heaps for the find... Looking into it...

dexX7 replied:
I think I found it:
```c++
                } else {
                    it++;
                }
```
... appears to be missing in the unsupported alert path in `CheckExpiredAlerts`.

zathras-crypto replied:
Shoot - you're right - nice find...  Yeah so if the alert hasn't expired the iterator isn't advanced and it'll just loop the same one endlessly - good catch mate thank you :)

zathras-crypto replied:
Fixed in faf84177d5d6d8eab2db3fb8aecf0cf4f5a5dd9c

dexX7 commented:
@zathras-crypto: what do you think is still missing, what do you want to add to finalize this PR?

dexX7 commented:
@zathras-crypto: what do you think is still missing, what do you want to add to finalize this PR?

zathras-crypto commented:
> @zathras-crypto: what do you think is still missing, what do you want to add to finalize this PR?

* Obtain & add foundation key
* Add a separate ```CheckActivationAuthorization()```
* Add a separate ```omniactivationallowsender```/```--omniactivationignoresender``` 
* Think on some test process and add some boost tests
* Do a final once-over

I've probably forgotten something, but it's all pretty straight forward stuff as you can see (I'm actually looking forward to getting my hands dirty with the boost tests)...  I'll push again for the foundation key at the meeting tomorrow and swing back to this PR...

dexX7 commented:
Yeah, so the core logic is finished, right? Then I can start to take another look at it.

zathras-crypto commented:
> Yeah, so the core logic is finished, right?

Yeah I hadn't planned on changing the fundamentals unless we find a bug or problem in approach...

dexX7 commented:
This probably needs a refinement:

![bets](https://cloud.githubusercontent.com/assets/5836089/9504978/e70fda16-4c3f-11e5-8dda-9cb22994ffae.png)

The tests I currently have (see reference one above) work very well though, after minimal adjustments. :)

zathras-crypto commented:
I see, so if we change the activation height with a secondary message the alerting isn't accurate - makes sense for the duplicate unsure why you have "will go live" and "is live" at the same time though, could you forward me your test to recreate this one in regtest?

There shouldn't be a max number of alerts, let me look into that...

dexX7 commented:
My mistake, you're right. Looks like there are more than 4 slots.

Here is a little script to generate alerts. I'm activating feature 3 more than once:

```bash
./src/qt/omnicore-qt -regtest -omniactivationallowsender=any
```
```bash
./src/omnicore-cli setgenerate true 101
ADDR=$(./src/omnicore-cli getnewaddress)
./src/omnicore-cli sendtoaddress $ADDR 0.1
BLOCKS=$(./src/omnicore-cli getblockcount)
./src/omnicore-cli omni_sendactivation $ADDR 3 $(($BLOCKS + 6)) 999
./src/omnicore-cli omni_sendactivation $ADDR 3 $(($BLOCKS + 8)) 999
./src/omnicore-cli omni_sendactivation $ADDR 3 $(($BLOCKS + 7)) 999
./src/omnicore-cli omni_sendactivation $ADDR 3 $(($BLOCKS + 8)) 999
./src/omnicore-cli omni_sendactivation $ADDR 3 $(($BLOCKS + 8)) 999
./src/omnicore-cli omni_sendactivation $ADDR 3 $(($BLOCKS + 8)) 999
./src/omnicore-cli omni_sendactivation $ADDR 3 $(($BLOCKS + 10)) 999
./src/omnicore-cli setgenerate true 1
./src/omnicore-cli omni_sendactivation $ADDR 3 $(($BLOCKS + 7)) 999
./src/omnicore-cli setgenerate true 1
```

dexX7 commented:
Since activating a feature more than once simply replaces the activation height, I think we should just keep track of the most recent activations, and maybe use map<featureId, activationBlock> to track it. Not sure how well this fits into the current approach of considering activations as alerts though. Alternatively maybe there could be a check, when adding a new alert, whether there is already an alert entry for the feature.

zathras-crypto commented:
Thanks for the feedback dude, I've changed a few bits to address your feedback - what are your thoughts on https://github.com/zathras-crypto/omnicore/commit/85c6afd64bcb0ad15b2513c06f3293ee24748910

dexX7 commented:
First the things I believe might be improved:

- we probably should use locks for alerts and activations, because the UI, PRC and core has access
- initially I thought `rules.cpp` would serve as simple file with some parameters, but it got bigger and bigger due to the "is allowed" checks, the checkpoints, consenus hashes, etc.
- the alert and notification stuff is now scattered across `omnicore.cpp`, `notifications.cpp` and `rules.cpp`
- at least on the first glimpse `AddPendingActivation`, `DeletePendingActivation`, `CheckLiveActivations`, `PendingActivationCompleted`, `GetPendingActivations` and `GetCompletedActivations` only serve the UI, RPC, so I tend to believe they might be moved out of `rules.cpp`; `CheckActivationAuthorization` potentially, too
- just an idea: would it make sense to flatten the result of `"omni_getactivations"` and simply return an array with alert entries, which have an additional field `"status"`?

I don't consider improving the points as requirement for this PR though.

Those are minor points, and I really, really like the general concept and the idea of having `"omni_getactivations"`, which seems great for testing!

Speaking of: yesterday, when I created the first activation tests, I run into the following issue: activating features completely messes with the tests. This is unfortunally the natural consequence, given that activation directly affect the consensus rules, but just to give two examples:

- the test for the DEx FP and "over-offer" deactivation can only be executed once, but no more, because then the feature is already activated
- after the FP  and "over-offer" deactivation, some of the other DEx tests fail (in particular, those for "over-offers")

A really ugly, but probably straight forward, workaround could be to have a hidden PRC, which is only allowed in regtest mode, to "reset" the activations and bypass the grace period restrictions (or directly modify the heights of the consensus params). I'm wondering, if you may have an idea to tackle this? :)

zathras-crypto commented:
This is really great feedback thanks, leave it with me :)

> deactivation can only be executed once, but no more, because then the feature is already activated

This got me thinking - what would happen, with current code, if we were to:

Block 100:  Feature Class C activation message for block 110 approved sender, ```NULLDATA_BLOCK``` set to 110
Block 109:  Class C transaction included in block, rejected as tx block is < ```NULLDATA_BLOCK```
Block 110:  Class C goes live
Block 111:  Class C transaction included in block, accepted as tx block is >= ```NULLDATA_BLOCK```
Block 112:  Feature Class C activation message for block 121 approved sender, ```NULLDATA_BLOCK``` set to 121
Block 113:  Class C transaction included in block, rejected as tx block is < ```NULLDATA_BLOCK```

So effectively, we could disable features again by perpetually broadcasting messages every n weeks pushing back the live block.  This wasn't my intention, and I can fix with checks, but it opened the question in my mind:

We're doing a lot of work on *enabling* features - what about *disabling* features - is that desirable?  For example take MetaDEx, it would be the equivalent of a "trading halt" in more traditional exchanges.  Should the foundation have this capability? (I think just like activations, no single person should have this capability).  Just thinking out loud...

dexX7 commented:
> is that desirable?

That's really a good question. I can see how an "emergency halt" may be useful, but on the other hand, *ideally* activation heights are immutable. That's a direct conflict.

zathras-crypto commented:
>> is that desirable?

> That's really a good question. I can see how an "emergency halt" may be useful, but on the other hand, ideally activation heights are immutable. That's a direct conflict.

Yeah, I agree there - and I'm adding a simple check to activations that fails if the feature ```_BLOCK``` is already past (ie the feature is already active) - making it so they can't be changed by future activation messages once live...  

It's not necessarily the methodology (eg using activations or something else) more the concept of disabling stuff and whether that's crossing some kind of boundary with the decentralized nature of the project - I definitely think there is benefit there for harm reduction as a last resort, but I don't know if it's the *right* thing to do hehe...

zathras-crypto commented:
> might be improved

Does this https://github.com/zathras-crypto/omnicore/commit/e7176f63e88f0ff42e5190315a8ef132dedbe0cb help?

zathras-crypto commented:
We also need to reset the ```_BLOCK``` values when ```clear_all_state``` is called - is there a simple way to reuse the values in the CConsensusParmas constructors?

zathras-crypto commented:
> A really ugly, but probably straight forward, workaround could be to have a hidden PRC, which is only allowed in regtest mode, to "reset" the activations and bypass the grace period restrictions (or directly modify the heights of the consensus params). I'm wondering, if you may have an idea to tackle this?

We could do something like this yeah, maybe perhaps a startup param should be required as a secondary check ```--omnienableoffchainchanges``` (or similar) which then allows usage of hidden RPC calls to modify the state away from what's on-chain (for testing purposes) but I'm not sure what other risks that introduces...

Another option would be to just orphan the chain with the activation in it with a newer (longer) one without an activation in, as long as its >50 blocks that should clear the slate and permit setting up of another test (though balances wouldn't be persisted etc)

dexX7 commented:
> It's not necessarily the methodology (eg using activations or something else) more the concept of disabling stuff and whether that's crossing some kind of boundary with the decentralized nature of the project

Right now, in the worst case, our other option would be to fire a hard alert and shutdown everything, which doesn't imply "this is good", but just to set some context. I kinda have a bad feeling about it, because it's so powerful, and the activations are somewhat more acceptable due to the grace period, so I tend to avoid "giving us even more power" (e.g. with an emergency reset of specific activations).

> Does this zathras-crypto@e7176f6 help?

Looks great!

> We also need to reset the _BLOCK values when clear_all_state is called ...

Do we? I'm probably mistaken, but I think we had a discussion about something similar before in the context of reorganizations, and that it is necessary to restart the client. I'm not sure anymore right now.

We'd need an extra function I think, because we don't store the initial default values, and we don't have access to the underlying objects, but adding one wouldn't be a big deal. I'll look into it, if we need one.

> Another option would be to just orphan the chain with the activation in it with a newer (longer) one without an activation in, as long as its >50 blocks that should clear the slate and permit setting up of another test (though balances wouldn't be persisted etc)

This is a very interesting idea! If we clear the activations as part of `clear_all_state()`, then this should work always, or am I missing something right now? Why did you mention >50 blocks in particular?

dexX7 commented:
I tested 87085395268f04fb234f1ec28935eeeda1ad7315 and it works to some degree, but even if the params are resetted, the effect is only temporary, because after a restart, or more generally, after re-processing the historical activation transactions, the activations are restored.

I'm split here, because it feels like the whole activation/alert stuff gets more and more complex, and I tend to prefer simplicity over features, at least at the moment and for those very critical parts. In this specific context it's because we don't have a model to reset, revert, step back, step forward, ... state as we'd like to, and there is the file based persistence, the special reorg handling of the SP DB, in-memory state objects, etc. etc.

zathras-crypto commented:
>> We also need to reset the _BLOCK values when clear_all_state is called ...

>Do we? I'm probably mistaken, but I think we had a discussion about something similar before in the context of reorganizations, and that it is necessary to restart the client. I'm not sure anymore right now.

```clear_all_state``` wipes the state and sets it to a blank slate, and begins parsing from Exodus.  If the client is reset to a fully blank slate then that should include any activations.  They will be processed again during parsing at the relevant block they're included in.

> This is a very interesting idea! If we clear the activations as part of clear_all_state(), then this should work always, or am I missing something right now? Why did you mention >50 blocks in particular?

It would work, but only once we add a call in ```clear_all_state``` to reset the consensus params to the original unmodified values.  I mention >50 blocks in particular because that's how long a chain you need to invalidate to cause the client to wipe the state and reparse from the beginning (we only keep persistence for roll-back for the last 50 blocks).

> because after a restart, or more generally, after re-processing the historical activation transactions, the activations are restored.

```clear_all_state``` wipes levelDB, ensuring on a restart there will be no activations in levelDB to restore. 

> it works to some degree

I whipped up a regtest to describe what I was trying to get across :)

* Clone my branch 0.0.10-Z-RevisedAlertsWithFeatureActivation
* Cherry pick the fix for next_spid values (git cherry-pick 9f56a81bb3e09e1e5662b7591d072367508252b4)
* Cherry pick your reset for in-memory consensus params (git cherry-pick 87085395268f04fb234f1ec28935eeeda1ad7315)
* Modify the consensus params for REGTEST mode for MSC_METADEX_BLOCK from 0 to 999999
* Run the below reg test script

Hopefully it's fairly self explanatory, but gives an idea of what I was getting at when I said we could orphan the chain to reset activations.

Here's the output I get:
```
zathras@zathras-desktop:~/github/build/omnicore$ ./regtest/test_reissueactivation.sh 
prepare some testnet BTC (note generate 150+ blocks to ensure some are confirmed after rollback)
obtain an address to work with
fund the address with some testnet BTC for fees
obtain some MSC
create some SPT#3
send MetaDEx transaction - prior to activation so should be invalid
    "valid" : false,
send activation for MetaDEx C in +6 blocks
send another MetaDEx transaction - should still be invalid until we mine some blocks
    "valid" : false,
mine enough blocks to activate MetaDEx
send another MetaDEx transaction - should now be valid
    "valid" : true,
rolling back chain >50 blocks (clears state & forces full reparse)
re-fund the address with some testnet BTC for fees
re-obtain some MSC
re-create some SPT#3
send MetaDEx transaction - should be invalid again
    "valid" : false,
resend activation for MetaDEx in +7 blocks
send another MetaDEx transaction - should still be invalid
    "valid" : false,
mine some more blocks to reactivate MetaDEx
send MetaDEx transaction - should now be valid again
    "valid" : true,
```

This does highlight some other problems though, namely with completed activations and the UI messages after a rollback (can't say I engineered for this scenario hehe)...  It's not something I've thought through in detail, I just wanted to note the concept, and we can clean it up if we want to continue down this path...

Does that help explain my thinking at all?  Sorry if not hehe :)

Reg test script:
```
#!/bin/bash

printf "prepare some testnet BTC (note generate 150+ blocks to ensure some are confirmed after rollback)\n"
./src/omnicore-cli --regtest setgenerate true 160 >null

printf "obtain an address to work with\n"
ADDR=$(./src/omnicore-cli --regtest getnewaddress MSCAccount)

printf "fund the address with some testnet BTC for fees\n"
./src/omnicore-cli --regtest sendtoaddress $ADDR 10 >null
./src/omnicore-cli --regtest setgenerate true 1 >null

printf "obtain some MSC\n"
JSON="{\"moneyqMan7uh8FqdCA2BV5yZ8qVrc9ikLP\":5,\""$ADDR"\":4}"
./src/omnicore-cli --regtest sendmany MSCAccount $JSON >null
./src/omnicore-cli --regtest setgenerate true 1 >null

printf "create some SPT#3\n"
./src/omnicore-cli --regtest omni_sendissuancefixed $ADDR 1 1 0 "Z_TestCat" "Z_TestSubCat" "Z_TestProperty" "Z_TestURL" "Z_TestData" 1000 >null
./src/omnicore-cli --regtest setgenerate true 1 >null

printf "send MetaDEx transaction - prior to activation so should be invalid\n"
TXID=$(./src/omnicore-cli --regtest omni_sendtrade $ADDR 1 10 3 50)
./src/omnicore-cli --regtest setgenerate true 1 >null
./src/omnicore-cli --regtest omni_gettransaction $TXID | grep valid

printf "send activation for MetaDEx C in +6 blocks\n"
BLOCKS=$(./src/omnicore-cli --regtest getblockcount)
./src/omnicore-cli --regtest omni_sendactivation $ADDR 2 $(($BLOCKS + 6)) 999 >null
./src/omnicore-cli --regtest setgenerate true 1 >null

printf "send another MetaDEx transaction - should still be invalid until we mine some blocks\n"
TXID=$(./src/omnicore-cli --regtest omni_sendtrade $ADDR 1 10 3 50)
./src/omnicore-cli --regtest setgenerate true 1 >null
./src/omnicore-cli --regtest omni_gettransaction $TXID | grep valid

printf "mine enough blocks to activate MetaDEx\n"
./src/omnicore-cli --regtest setgenerate true 7 >null

printf "send another MetaDEx transaction - should now be valid\n"
TXID=$(./src/omnicore-cli --regtest omni_sendtrade $ADDR 1 10 3 50)
./src/omnicore-cli --regtest setgenerate true 1 >null
./src/omnicore-cli --regtest omni_gettransaction $TXID | grep valid

printf "rolling back chain >50 blocks (clears state & forces full reparse)\n"
BLOCKS=$(./src/omnicore-cli --regtest getblockcount)
BLOCKHASH=$(./src/omnicore-cli --regtest getblockhash $(($BLOCKS - 60)))
./src/omnicore-cli --regtest invalidateblock $BLOCKHASH >null
./src/omnicore-cli --regtest setgenerate true 1 >null
BLOCKS=$(./src/omnicore-cli --regtest getblockcount)

printf "re-fund the address with some testnet BTC for fees\n"
./src/omnicore-cli --regtest sendtoaddress $ADDR 10 >null
./src/omnicore-cli --regtest setgenerate true 1 >null

printf "re-obtain some MSC\n"
JSON="{\"moneyqMan7uh8FqdCA2BV5yZ8qVrc9ikLP\":5,\""$ADDR"\":4}"
./src/omnicore-cli --regtest sendmany MSCAccount $JSON >null
./src/omnicore-cli --regtest setgenerate true 1 >null

printf "re-create some SPT#3\n"
./src/omnicore-cli --regtest omni_sendissuancefixed $ADDR 1 1 0 "Z_TestCat" "Z_TestSubCat" "Z_TestProperty" "Z_TestURL" "Z_TestData" 1000 >null
./src/omnicore-cli --regtest setgenerate true 1 >null

printf "send MetaDEx transaction - should be invalid again\n"
TXID=$(./src/omnicore-cli --regtest omni_sendtrade $ADDR 1 10 3 50)
./src/omnicore-cli --regtest setgenerate true 1 >null
./src/omnicore-cli --regtest omni_gettransaction $TXID | grep valid

printf "resend activation for MetaDEx in +7 blocks\n"
BLOCKS=$(./src/omnicore-cli --regtest getblockcount)
./src/omnicore-cli --regtest omni_sendactivation $ADDR 2 $(($BLOCKS + 7)) 999 >null

printf "send another MetaDEx transaction - should still be invalid\n"
TXID=$(./src/omnicore-cli --regtest omni_sendtrade $ADDR 1 10 3 50)
./src/omnicore-cli --regtest setgenerate true 1 >null
./src/omnicore-cli --regtest omni_gettransaction $TXID | grep valid

printf "mine some more blocks to reactivate MetaDEx\n"
./src/omnicore-cli --regtest setgenerate true 9 >null

printf "send MetaDEx transaction - should now be valid again\n"
TXID=$(./src/omnicore-cli --regtest omni_sendtrade $ADDR 1 10 3 50)
./src/omnicore-cli --regtest setgenerate true 1 >null
./src/omnicore-cli --regtest omni_gettransaction $TXID | grep valid
```

dexX7 commented:
Thanks, great information and great test!

It's very similar to the test I prepared for the [DEx feature](https://github.com/dexX7/bitcoin-spock/blob/test-feature-activation/omnij-rpc/src/integ/groovy/foundation/omni/test/rpc/activation/OverOfferDeactivationSpec.groovy), and if I find some time, I'm going to adopt it for OmniJ! :)

> reset for in-memory consensus params

Alright, I agree, we should clear the params via `clear_all_state`, but get rid of `omni_resetparams`.

I think my main hesitation was due to the mix with the RPC to clear the state externally, which feels pretty hacky and unreliable.

Regarding the tests: mining 50 extra blocks, only to wipe the state, is also very hacky, especially since 50 blocks persistence is very arbitrary, but I'm going to try it later, and I prefer it over the RPC, because it works not only temporary, but is reliable.

When invalidating blocks, I noticed there is sometimes the issue that there are some transactions in the mempool, and we have a hidden call `clearmempool` to handle it. I usually go this route: `invalidateblock -> clearmempool -> set generate 1`.

zathras-crypto commented:
Rebased (I think! hehe)

dexX7 commented:
I suggest we adopt 87085395268f04fb234f1ec28935eeeda1ad7315 (but without RPC) to reset the params.

Once adopted, what else do you want to add? (if there is something else) Just trying to get an understanding how far we are away from getting this in. :)

zathras-crypto commented:
Paging @CraigSellars - any progress with the foundation P2SH key for activating features?

dexX7 commented on src/omnicore/notifications.h:
```diff
@@ -3,26 +3,48 @@
+#include <vector>
+
+//! Alert types
+#define ALERT_FEATURE_UNSUPPORTED     0 // not allowed via alert broadcast
```
Was there a specific reason to use `#define` instead of `const uint16_t`? Just thinking, because `uint16_t` is also the native type used for the messages.

zathras-crypto replied:
Hmm, nope - I'm not really sure on best practices for programming :(  I just knew the compiler would replace those definitions with the values and thus it made the code easier to read - but if it's better practice to use ```const``` with the native type instead I can certainly start doing that :)

dexX7 replied:
Yeah, you can probably make a case for both sides (or `enum`). The upside of `const T` is that it's contained by namespaces, which can be desired. For example, we could have something like:
```c++
namespace omnicore {
namespace rules {
const unsigned int max_multisig_payloads = 2;
const unsigned int max_sequence_number = 255;
}
}
```
I have no strong opinion about this though. :)

dexX7 commented on src/omnicore/rules.cpp:
```diff
@@ -322,48 +338,66 @@ bool ActivateFeature(uint16_t featureId, int activationBlock, int transactionBlo
+        std::string alertText = strprintf("Your client must be updated and will shutdown at block %d (unsupported feature %d ('%s') activated)\n",
+                                          activationBlock, featureId, featureName);
+        AddAlert("omnicore", ALERT_BLOCK_EXPIRY, activationBlock, alertText);
+        CAlert::Notify(alertText, true);
```
Hmm. We fire the other notification in `logicMath`. I'm wondering, whether we could/should move `CAlert::Notify` into `AddAlert`, ideally including the log output. Would this make sense? If so, is it possible, or would this trigger notifications, which shouldn't be triggered? (e.g. when loading alerts or so)

zathras-crypto replied:
All alerts from levelDB are 'added' and then immediately expired (if appropriate) when starting the client so if we moved it into ```AddAlert``` we would be triggering ```CAlert::Notify``` for old alerts during startup.

dexX7 commented on src/omnicore/test/alert_tests.cpp:
```diff
@@ -79,7 +79,7 @@ BOOST_AUTO_TEST_CASE(alert_authorize_any_source)
     mapArgs = mapArgsOriginal;
 }
-
+/*
```
Unless this is only temporary, I think we may just remove commented code completely.

dexX7 commented on src/omnicore/rpctx.cpp:
```diff
@@ -1085,14 +1085,20 @@ Value omni_sendalert(const Array& params, bool fHelp)
+    }
+    uint16_t alertType = static_cast<uint16_t>(tempAlertType);
+    int64_t tempExpiryValue = params[2].get_int64();
+    if (tempExpiryValue < 1 || 4294967295LL < tempExpiryValue) {
```
With the [current version schema](https://github.com/OmniLayer/omnicore/pull/162/files#diff-da586973b290a2920a045b8ad5ddf4b9R47) we shall never reach 1.0.0.0, hehe. ;)

dexX7 commented on src/rpcserver.cpp:
```diff
@@ -357,23 +357,25 @@ static const CRPCCommand vRPCCommands[] =
 
     /* Omni Core data retrieval calls */
     /* CATEGORY                              NAME                               ACTOR (FUNCTION)                  OKSAFEMODE  THREADSAFE  REQWALLET */
-    { "omni layer (data retrieval)",         "omni_getinfo",                    &omni_getinfo,                    true,       true,       false },
```
Uh, oh!

Looks like this is the old version (probably introduced when resolving conflicts).

dexX7 commented on src/omnicore/rpctx.cpp:
```diff
@@ -1021,7 +1021,7 @@ Value omni_sendchangeissuer(const Array& params, bool fHelp)
-    if (fHelp || params.size() != 3)
+    if (fHelp || params.size() != 4)
         throw runtime_error(
             "omni_sendactivation \"fromaddress\" featureid block\n"
```
`minclientversion` missing

dexX7 commented:
The changes in the RPC server should be reverted to the current tip, see: https://github.com/OmniLayer/omnicore/pull/176#discussion_r38371307. :)

I tested the updated version, and it works as expected!

I stepped through the OmniJ tests to observe `"omni_getactivations"` and the UI, and the alerts are properly replaced. I also confirmed that already confirmed alerts can't be overwritten, and finally that re-schedueled alerts are still ordered by height, e.g. in the following case:

```
[block 103] activate feature 3 at block 108 (pending: [3 -> 108])
[block 104] activate feature 5 at block 109 (pending: [3 -> 108, 5 -> 109])
[block 104] activate feature 3 at block 110 (pending: [5 -> 109, 3 -> 110])
```

dexX7 commented:
@zathras-crypto: would you consider this PR as ready?

zathras-crypto commented:
Rebased on current tip.

dexX7 commented:
I added a bunch of new regtests for the feature activation. If you're interested, see here:

- https://github.com/OmniLayer/OmniJ/pull/98/files

It's probably clear what happens during the tests, but feel free to ask, if you'd like to learn more about it. :)

Unfortunally the tests are currently not active, because:

- the activation tests can only be run once, because otherwise the features were already activated
- the traditional DEx activation tests collide with two other traditional DEx tests
- the MetaDEx tests only work, if the MetaDEx is initially deactivated, which is not the case, even in regtest mode
- if we deactivate the MetaDEx in regtest mode (which is very do-able), then the standard MetaDEx tests may fail, in case the tests are run before the activation

Not sure, how to resolve this ... anyway, I manually run the tests after disabling the MetaDEx in regtest mode, and FWIW: they passed with this PR. :)

dexX7 commented:
Ha! Alright, I tested your suggestion! :)

- at the beginning of the test bundle, mine one block and store it's hash
- at the end of the test bundle, mine 50 blocks
- then invalidate the initial block
- clear mempool, mine another block to finalize

While this doesn't resolve the MetaDEx issue, it can be used for the grace period and tradtional DEx tests.

dexX7 commented:
@zathras-crypto: please add `-omniactivationallowsender=any` to the test script `qa/pull-tester/omnicore-rpc-tests.sh`:

```diff
diff --git a/qa/pull-tester/omnicore-rpc-tests.sh b/qa/pull-tester/omnicore-rpc-tests.sh
index cdb802a..556994d 100755
--- a/qa/pull-tester/omnicore-rpc-tests.sh
+++ b/qa/pull-tester/omnicore-rpc-tests.sh
@@ -16,7 +16,7 @@ touch "$DATADIR/regtest/omnicore.log"
 cd $TESTDIR
 echo "Omni Core RPC test dir: "$TESTDIR
 echo "Last OmniJ commit: "$(git log -n 1 --format="%H Author: %cn <%ce>")
-$REAL_BITCOIND -regtest -txindex -server -daemon -rpcuser=bitcoinrpc -rpcpassword=pass -debug=1 -omnidebug=all -omnialertallowsender=any -discover=0 -listen=0 -datadir="$DATADIR" &
+$REAL_BITCOIND -regtest -txindex -server -daemon -rpcuser=bitcoinrpc -rpcpassword=pass -debug=1 -omnidebug=all -omnialertallowsender=any -omniactivationallowsender=any -discover=0 -listen=0 -datadir="$DATADIR" &
 $REAL_BITCOINCLI -regtest -rpcuser=bitcoinrpc -rpcpassword=pass -rpcwait getinfo
 $REAL_BITCOINCLI -regtest -rpcuser=bitcoinrpc -rpcpassword=pass -rpcwait getinfo_MP
 ./gradlew --console plain :omnij-rpc:regTest
```

zathras-crypto commented:
> I added a bunch of new regtests for the feature activation

This is fantastic, thanks dude!

> if we deactivate the MetaDEx in regtest mode (which is very do-able), then the standard MetaDEx tests may fail, in case the tests are run before the activation

Do the OmniJ tests run in sequence?  Is it feasible to structure them so that activation tests run before MetaDEx tests?

dexX7 commented:
> Do the OmniJ tests run in sequence?

No, usually each test is atomic. There are "stepwise tests", which are basically a sequence of tests, but the whole test suite is at best a loose collection of tests or test sequences, whereby no order is enforced (and I'm unfortunally not sure, if it's possible to enforce one). This worked nicely so far, and it allows to run tests without clearing the datadir etc., but here it's really not useful.

As alternative I could think about a workaround where the MetaDEx activation is embedded into the regular MetaDEx tests, too, which are ultimately reverted.

dexX7 commented:
@zathras-crypto:

I haven't seen much progress here, so I'm wondering what causes the stalling?

As far as I can see only the activation senders are missing, which we might add later.

I would really like to get this in rather sooner or later, even if it's incomplete.

> Do the OmniJ tests run in sequence?

I partially resolved this via https://github.com/OmniLayer/OmniJ/pull/108, which runs the activation tests seperately before the actual regtests.

All the activation tests are currently disabled though, since they are depending on this PR. :)

zathras-crypto commented:
I've rebased this PR against the current tip and cleaned up conflicts plus a little more testing.  I think this is good to go if you're happy with it @dexX7 ?

Also I saw on email we got the final required key for the P2SH hard coded activation address, so think we have all the pieces we need to merge in activations?

Thanks for all your help (and patience!) :)
Z

zathras-crypto commented:
```
Offering more than available on the distributed exchange is still valid until the feature activation FAILED
    groovy.lang.MissingMethodException: No signature of method: foundation.omni.test.rpc.activation.OverOfferDeactivationSpec.createDexSellOffer() is applicable for argument types: (org.bitcoinj.core.Address, foundation.omni.CurrencyID, java.math.BigDecimal, java.math.BigDecimal, java.lang.Byte, org.bitcoinj.core.Coin, java.lang.Byte) values: [mxCoKLvkUMoJxWLDPp49aMXNphHb9YyCym, CurrencyID:1, 0.50000000, ...]
    Possible solutions: createDexSellOffer(org.bitcoinj.core.Address, foundation.omni.CurrencyID, foundation.omni.OmniDivisibleValue, org.bitcoinj.core.Coin, java.lang.Byte, org.bitcoinj.core.Coin, java.lang.Byte)
        at foundation.omni.test.rpc.activation.OverOfferDeactivationSpec.Offering more than available on the distributed exchange is still valid until the feature activation(OverOfferDeactivationSpec.groovy:76)
    After the successful activation of the feature, it is no longer valid to offer more than available
    The activation has no effect on orders, which offer exactly the balance that is available

18 tests completed, 1 failed, 2 skipped
:omnij-rpc:activationRegTest FAILED
```

dexX7 commented:
> :omnij-rpc:activationRegTest FAILED

Argh, looks like a recent update of OmniJ broke the tests.

I'm going to tackle it later!

> Also I saw on email we got the final required key for the P2SH hard coded activation address, so think we have all the pieces we need to merge in activations?

I think so, too! :)

I quickly updated the doc:
https://gist.github.com/dexX7/84a7d327567d1c1e30f7#file-3_redeemscript-md

dexX7 commented:
Damn! I didn't notice you pushed the activation key already.

Otherwise I would have merged this PR, and then #142. :(

I'm very close to reverting #142 (partially), but this would probably be even messier.

The conflict is caused by:

- https://github.com/dexX7/bitcoin/compare/523f79071128b3a6868a5131f0cb205a00ef04a6...dexX7:oc-0.10-avoid-activation-conflict

I'm sorry for the inconvenience, but can you please resolve?

I probably wouldn't have noticed a label update either, so for the future, I suggest a pull request that changes from "in progress" to "ready" should also come with a comment, to trigger GitHub's notifications.

But a big note to myself: double check all submissions, which are close to be merged.

zathras-crypto commented:
> Damn! I didn't notice you pushed the activation key already.

Sorry, I should have posted but held off as I was planning to setup a test transaction for 1 TMSC from this address & have us all test but just hadn't got round to it yet :)

> I'm sorry for the inconvenience, but can you please resolve?

Not an inconvenience at all dude!!!  I've rebased on the current tip, so should be OK now :)
